micrometer-observation/src/main/java/io/micrometer/observation/aop/ObservedAspect.java

```diff
@@ -0,0 +1,260 @@
+                throw error;
+            }
+        }
+        else {
+            return observation.observeChecked(() -> pjp.proceed());
+        }
+    }
+
+    private Class<?> getDeclaringClass(ProceedingJoinPoint pjp) {
+        Method method = ((MethodSignature) pjp.getSignature()).getMethod();
+        Class<?> declaringClass = method.getDeclaringClass();
+        if (!declaringClass.isAnnotationPresent(Observed.class)) {
+            return pjp.getTarget().getClass();
+        }
+
+        return declaringClass;
+    }
+
+    private Method getMethod(ProceedingJoinPoint pjp) throws NoSuchMethodException {
+        Method method = ((MethodSignature) pjp.getSignature()).getMethod();
+        if (method.getAnnotation(Observed.class) == null) {
+            return pjp.getTarget().getClass().getMethod(method.getName(), method.getParameterTypes());
+        }
+
+        return method;
+    }
+
+    private void stopObservation(Observation observation, Observation.Scope scope, @Nullable Throwable error) {
+        if (error != null) {
+            observation.error(error);
+        }
+        scope.close();
+        observation.stop();
+    }
+
+    public enum ObservedAspectObservation implements DocumentedObservation {
+
+        DEFAULT;
+
+        static Observation of(ProceedingJoinPoint pjp, Method method, Observed observed, ObservationRegistry registry,
+                @Nullable Observation.KeyValuesProvider<ObservedAspectContext> keyValuesProvider) {
+            String name = observed.name().isEmpty() ? DEFAULT_OBSERVATION_NAME : observed.name();
+            Signature signature = pjp.getStaticPart().getSignature();
+            String contextualName = observed.contextualName().isEmpty()
+                    ? signature.getDeclaringType().getSimpleName() + "#" + signature.getName()
+                    : observed.contextualName();
+
+            Observation observation = Observation.createNotStarted(name, new ObservedAspectContext(pjp), registry)
+                    .contextualName(contextualName)
+                    .lowCardinalityKeyValue(CLASS_NAME.getKeyName(), signature.getDeclaringTypeName())
+                    .lowCardinalityKeyValue(METHOD_NAME.getKeyName(), signature.getName())
+                    .lowCardinalityKeyValues(KeyValues.of(observed.lowCardinalityKeyValues()));
+
+            if (keyValuesProvider != null) {
+                observation.keyValuesProvider(keyValuesProvider);
+            }
+
+            return observation;
+        }
+
+        @Override
+        public String getName() {
+            return "%s";
+        }
+
+        @Override
+        public String getContextualName() {
+            return "%s";
+        }
+
+        @Override
+        public KeyName[] getLowCardinalityKeyNames() {
+            return ObservedAspectLowCardinalityKeyName.values();
+        }
+
+        public enum ObservedAspectLowCardinalityKeyName implements KeyName {
```

Comment:
If this is not set as an enum in a separate file we will have a problem with micrometer docs generator

Reply:
You mean `ObservedAspectObservation` should be in a separate file or both `ObservedAspectObservation` and `ObservedAspectLowCardinalityKeyName`?

I'm asking because `DocumentedObservation` says:
> If the observation contains {@link KeyName} then those need to be declared as nested enums

Reply:
The Observation should be in a separate file and it should have a nested enum with keys

Reply:
fixed in https://github.com/micrometer-metrics/micrometer/commit/86befc188c7da809fd5e5c5ca6c106601797f435
Thank you!